### PR TITLE
ctc_loss

### LINKED ISSUE #12745: ctc_loss

### PR DESCRIPTION
**Pull request by Cemilcelik11**
Close #12745

### CODE REVIEW

**Comment by Cemilcelik11** [comment]
Hello, I was wondering if there's any development on this PR. Thank you!

**Comment by Cemilcelik11** [comment]
Close #12745

**Comment by Cemilcelik11** [comment]
Close #12745

**Comment by Cemilcelik11** [comment]
I was also wondering whether I could move on to the next phase of my Ivy Machine Learning Intern application since I have done the task of Pytorch frontend. Of course I will continue with implementing this function for the backend too but if there is no error in frontend implementation can I move to the next phase of my application?

**Comment by Cemilcelik11** [comment]
I think I missed the fact: You already mentioned the Extend ıvy Api task and when I found that task I saw I have to do the implementation for both ivy api and the backend api. I think that is why it fails. now I am gonna concentrate on that.

**Comment by Cemilcelik11** [comment]
I added code for backend api and ivy api and a test function in test_ivy/test_functional/test_experimental/test_nn/test_losses.py
The codes for backend and ivy are also to find in experimental directories.

File: ivy/functional/ivy/experimental/losses.py
Comment (YushaArif99) [inline ivy/functional/ivy/experimental/losses.py:1]: You've added a compositional implementation for `ivy.ctc_loss`. Any particular reason for doing so? Like do all other backends not natively support this function? If so, then this implementation makes sense. But if all other backends do contain the `ctc_loss` function, then you should simply defer to the backend implementation like so: 
```
return current_backend(*args).ctc_loss(*args, **kwargs)
```

File: ivy/functional/backends/torch/experimental/losses.py
Comment (YushaArif99) [inline ivy/functional/backends/torch/experimental/losses.py:1]: Could you also kindly add the implementations for the other backends similarly 🙂

File: ivy_tests/test_ivy/test_frontends/test_torch/.ivy/key.pem
Comment (YushaArif99) [inline ivy_tests/test_ivy/test_frontends/test_torch/.ivy/key.pem:1]: Could you kindly remove this file from the modified files 🙂

File: ivy_tests/test_array_api
Comment (YushaArif99) [inline ivy_tests/test_array_api:1]: Same goes for this file as well

**Review by YushaArif99 — CHANGES_REQUESTED** [review]
Hi @Cemilcelik11 

Thanks for adding the backend implementation for the Torch backend. Could you kindly also address the abovementioned points and re-request a review? 

As always, let me know if you have any other questions 🙂

**Comment by YushaArif99** [comment]
> I was also wondering whether I could move on to the next phase of my Ivy Machine Learning Intern application since I have done the task of Pytorch frontend. Of course I will continue with implementing this function for the backend too but if there is no error in frontend implementation can I move to the next phase of my application?


Unfortunately, that's not for me to decide 😅. I think it's better to PM `@IvyTeam` on Discord as they are the ones that handle the queries related to the internship application. 

Though, the only issue I would say is that without actually running the tests, it's kind of difficult to say whether the implementation is correct or not. Hope that makes sense 🙂

File: ivy/functional/ivy/experimental/losses.py
Comment (Cemilcelik11) [reply ivy/functional/ivy/experimental/losses.py:1]: thank you, right now working on it!!

File: ivy/functional/backends/torch/experimental/losses.py
Comment (Cemilcelik11) [reply ivy/functional/backends/torch/experimental/losses.py:1]: sure!

**Comment by Cemilcelik11** [comment]
Sorry to bother you guys, couldn't understand why it fails.

**Comment by Cemilcelik11** [comment]
For my understanding, I had to implement this function for all the backends like torch, tensorflow etc. That is what I did.

File: ivy/functional/backends/jax/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/backends/jax/experimental/losses.py:4]: `cannot import name 'stax' from 'jax.experimental'`

File: ivy/functional/backends/tensorflow/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/backends/tensorflow/experimental/losses.py:35]: in a backend implementation, you should only use functions of that backend only, also there is no such function

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]

**Comment by Cemilcelik11** [comment]
I made sure to use functions of that specific backend only, eliminated some errors but when I run the two test files(frontend test and test_functional/test_experimental/test_nn/test.losses.py) I get IvyBackendException, couldn't solve for several hours so I committed.  Error: fatal: No url found for submodule path 'ivy_tests/test_array_api' in .gitmodules
  Error: The process '/usr/bin/git' failed with exit code 128
  that is the error message I see on github now.

File: ivy/functional/ivy/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/ivy/experimental/losses.py:174]: you are not passing the rest of the keyword arguments
```suggestion
    return current_backend(log_probs, targets, input_lengths, target_lengths).ctc_loss(
    log_probs,
    targets,
    input_lengths,
    target_lengths,
    axis=axis,
    blank=blank,
    zero_infinity=zero_infinity,
    reduction=reduction,
    out=out,
)
```

File: ivy_tests/test_ivy/test_frontends/test_torch/.ivy/key.pem
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_frontends/test_torch/.ivy/key.pem:1]: delete this

File: ivy/functional/frontends/torch/nn/functional/loss_functions.py
Comment (MahmoudAshraf97) [inline ivy/functional/frontends/torch/nn/functional/loss_functions.py:564]: no need for the `try except`
```suggestion

    ret = ivy.ctc_loss(log_probs, labels, input_length, label_length, blank=blank, zero_infinity=zero_infinity)
    ret = _apply_reduction(reduction, ret)
    return ret
```

File: ivy_tests/test_array_api
Comment (MahmoudAshraf97) [inline ivy_tests/test_array_api:1]: delete this

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py
Comment (MahmoudAshraf97) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:145]: pass the `np.int32` as a string instead
```suggestion
        input_dtypes=dtype_and_logits[0] + dtype_and_targets[0] + logit_lengths[0] + label_lengths[0] +"int32",
```

File: ivy/functional/backends/paddle/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/backends/paddle/experimental/losses.py:14]: add the `@with_unsupported_device_and_dtypes` handler as this function doesn't support all dtypes

File: ivy/functional/backends/jax/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/backends/jax/experimental/losses.py:20]:
```suggestion

```

File: ivy/functional/backends/jax/experimental/losses.py
Comment (MahmoudAshraf97) [inline ivy/functional/backends/jax/experimental/losses.py:28]: import this from jax itself not ivy jax

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]
before merging I'd like to request @CatB1t review for the tests

**Comment by Cemilcelik11** [comment]
Thanks for the quick response! I am applyiing the changes. Quick question: do I need to run the command git reset –hard upstream/master every time I want to make a commit or git fetch upstream is enough??

File: ivy/functional/backends/jax/experimental/losses.py
Comment (Cemilcelik11) [reply ivy/functional/backends/jax/experimental/losses.py:28]: when I look at other codes I saw people imported like this and when I hover over JaxArray pycharm shows me it belongs to ivy.functional.backend.jax

Comment (Cemilcelik11) [reply ivy/functional/backends/jax/experimental/losses.py:28]: do you want me to use jax.numpy.array?

Comment (Cemilcelik11) [reply ivy/functional/backends/jax/experimental/losses.py:28]: andI can't import  JAxArray from jax directly.

Comment (Cemilcelik11) [reply ivy/functional/backends/jax/experimental/losses.py:28]: jax.numpy.ndarray
I am gonna try this

**Review by MahmoudAshraf97 — CHANGES_REQUESTED** [review]
please make sure the arguments are the same and unified between the 5 backends and Ivy, also the output and return format must be the same

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py
Comment (Cemilcelik11) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py]: but then I get the error: TypeError: can only concatenate list (not "str") to list

Comment (Cemilcelik11) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py]: I think it is because of the helpers.dtype_and_values which returns:
A strategy that draws a tuple of a list of dtypes and a list of their respective arrays.

Comment (Cemilcelik11) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py]: maybe [np.int32] was right at the first place, what do you think?

Comment (Cemilcelik11) [reply ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py]: and is there a way to push the changes without this command:
git reset --hard upstream/master
because of this I add the codes manually all over again every time I make a commit:)

**Pull request closed without merging**
